modules/build/src/main/scala/scala/build/preprocessing/mdsandbox/SnippetPackager.scala

```diff
@@ -0,0 +1,147 @@
+package scala.build.preprocessing.mdsandbox
+
+import MarkdownSnippet.Fence
+
+class SnippetPackager(val fileName: String, val snippets: Seq[Fence]) {
+  def this(fileName: String, content: String) = this(fileName, MarkdownSnippet.findFences(content))
+
+  val (rawSnippets, processedSnippets) = snippets.partition(f => f.isRaw)
+  val (testSnippets, runSnippets) = processedSnippets.partition(f => f.isTest)
+
+  val runObjectIdentifier: String = s"`Markdown$$${fileName.replace('.', '$')}`"
+  val testObjectIdentifier: String = s"`Markdown_Test$$${fileName.replace('.', '$')}`"
+
+  /** Generates class name for a snippet with given index */
+  def runClassName(index: Int): String = s"Snippet$$$index"
+
+  /** Generates class name for a test snippet with given index */
+  def testClassName(index: Int): String = s"`Test$$${fileName.replace('.', '$')}_$index`"
+
+  /** Returns Scala snippets packed into classes and glued together into an object */
+  def buildScalaMain(): String = {
+    if (runSnippets.isEmpty) s"object $runObjectIdentifier {def execute(): Unit = {}}"  // no snippets
+    else (0 until runSnippets.length).foldLeft(
+      s"object $runObjectIdentifier {def execute(): Unit = {"
+    ) (
+      (sum, index) => 
+        if (runSnippets(index).resetScope || index == 0) sum :++ s"new ${runClassName(index)}; "
+        else sum  // that class hasn't been created
+    )
+    .:++("} ")
+    .:++(buildScalaMain(0, 0))
+    .:++("}")
+  }
+
+  private def buildScalaMain(index: Int, line: Int): String = {
+    if (index >= runSnippets.length) "}"  // close last class
+    else {
+      val fence: Fence = runSnippets(index)
+      val classOpener: String =
+        if (index == 0)            s"class ${runClassName(index)} {\n"     // first snippet needs to open a class
+        else if (fence.resetScope) s"}; class ${runClassName(index)} {\n"  // if scope is being reset, close previous class and open a new one
+        else "\n"
+      ("\n" * (fence.startLine - line - 1))                 // padding
+        .:++(classOpener)                                   // new class opening (if applicable)
+        .:++(fence.body)                                    // snippet body
+        .:++("\n")                                          // padding in place of closing backticks
+        .:++(buildScalaMain(index + 1, fence.endLine + 1))  // further snippets
+    }
+  }
+
+  /** Returns test snippets packed into classes and glued together into an object */
+  def buildScalaTest(): String = {
+    if (testSnippets.isEmpty) s"object $testObjectIdentifier {}"  // no snippets
+    else buildScalaTest(0, 0) :++ "}"
+  }
+
+  /*
+   * TESTING APPROACH A - RAW TESTING
```

Comment:
seems like we'll be going with approach A, so the code needs to solidify on that